{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Import\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "import fuzzysearch\n",
    "import matplotlib as plt\n",
    "plt.rcParams['pdf.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Load key of conditions and files\n",
    "samples = pd.read_excel(\"Edit_site_analysis_demo_key.xls\")[[\"Phage\",\"Gene\",\"Plasmid\",\"Direction\",\"edit name\",\"Genome Position\",\"WT nt\",\"Edited nt\",\"L_inside\",\"R_inside\",\"L_outside\",\"R_outside\",\"Rep1\",\"Rep2\",\"Rep3\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phage</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Plasmid</th>\n",
       "      <th>Direction</th>\n",
       "      <th>edit name</th>\n",
       "      <th>Genome Position</th>\n",
       "      <th>WT nt</th>\n",
       "      <th>Edited nt</th>\n",
       "      <th>L_inside</th>\n",
       "      <th>R_inside</th>\n",
       "      <th>L_outside</th>\n",
       "      <th>R_outside</th>\n",
       "      <th>Rep1</th>\n",
       "      <th>Rep2</th>\n",
       "      <th>Rep3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lambda</td>\n",
       "      <td>L</td>\n",
       "      <td>pSBK.147</td>\n",
       "      <td>F</td>\n",
       "      <td>L_TAA_TGA</td>\n",
       "      <td>14126</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>CTTTCGCAGT</td>\n",
       "      <td>AATCCCATGA</td>\n",
       "      <td>TGTAAGTTCCGCAATAACGT</td>\n",
       "      <td>CGATGTGCGCCAGCGGAGTC</td>\n",
       "      <td>msSBK_34_01</td>\n",
       "      <td>msSBK_34_25</td>\n",
       "      <td>msCF_02_134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Lambda</td>\n",
       "      <td>L</td>\n",
       "      <td>pSBK.148</td>\n",
       "      <td>R</td>\n",
       "      <td>L_TAA_TGA</td>\n",
       "      <td>14126</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>CTTTCGCAGT</td>\n",
       "      <td>AATCCCATGA</td>\n",
       "      <td>TGTAAGTTCCGCAATAACGT</td>\n",
       "      <td>CGATGTGCGCCAGCGGAGTC</td>\n",
       "      <td>msSBK_34_02</td>\n",
       "      <td>msSBK_34_26</td>\n",
       "      <td>msCF_02_135</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Phage Gene   Plasmid Direction  edit name  Genome Position WT nt  \\\n",
       "0  Lambda    L  pSBK.147         F  L_TAA_TGA            14126     A   \n",
       "1  Lambda    L  pSBK.148         R  L_TAA_TGA            14126     A   \n",
       "\n",
       "  Edited nt    L_inside    R_inside             L_outside  \\\n",
       "0         G  CTTTCGCAGT  AATCCCATGA  TGTAAGTTCCGCAATAACGT   \n",
       "1         G  CTTTCGCAGT  AATCCCATGA  TGTAAGTTCCGCAATAACGT   \n",
       "\n",
       "              R_outside         Rep1         Rep2         Rep3  \n",
       "0  CGATGTGCGCCAGCGGAGTC  msSBK_34_01  msSBK_34_25  msCF_02_134  \n",
       "1  CGATGTGCGCCAGCGGAGTC  msSBK_34_02  msSBK_34_26  msCF_02_135  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Globals\n",
    "outcomes_dict = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Defs\n",
    "def extract_and_match(read,index,rep):\n",
    "    left_outside = fuzzysearch.find_near_matches(samples.iloc[index][\"L_outside\"],read,max_l_dist=2)\n",
    "    right_outside = fuzzysearch.find_near_matches(samples.iloc[index][\"R_outside\"],read,max_l_dist=2)\n",
    "    if len(left_outside) == 1 and len(right_outside) == 1:\n",
    "        left_inside = fuzzysearch.find_near_matches(samples.iloc[index][\"L_inside\"],read,max_l_dist=1)\n",
    "        right_inside = fuzzysearch.find_near_matches(samples.iloc[index][\"R_inside\"],read,max_l_dist=1)\n",
    "        if len(left_inside) == 1 and len(right_inside) == 1:\n",
    "            var_nt = read[left_inside[0].end:right_inside[0].start]\n",
    "            if var_nt == samples.iloc[index][\"WT nt\"]:\n",
    "                return 'wt'\n",
    "            elif var_nt == samples.iloc[index][\"Edited nt\"]:\n",
    "                return 'edited'\n",
    "            else:\n",
    "                return 'unmatched_edit_nt'\n",
    "        else:\n",
    "            return 'unmatched_region'\n",
    "    else:\n",
    "        return 'unmatched_region'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "msSBK_34_01\n",
      "msSBK_34_25\n",
      "msCF_02_134\n",
      "msSBK_34_02\n",
      "msSBK_34_26\n",
      "msCF_02_135\n"
     ]
    }
   ],
   "source": [
    "#step through samples\n",
    "for i in samples.index:\n",
    "    sample_i = int(i)\n",
    "    outcomes_dict[i] = {}\n",
    "    for rep in [\"Rep1\",\"Rep2\",\"Rep3\"]:\n",
    "        outcomes_dict[i][rep]= {'wt':0,'edited':0,'unmatched_region':0,'unmatched_edit_nt':0}\n",
    "        all_reads_str = []\n",
    "        read_counter = []\n",
    "        fastq_reads = \"./%s.fastq\" % samples[rep][i]\n",
    "        try:\n",
    "            for seq_record in SeqIO.parse(fastq_reads, \"fastq\"):\n",
    "                all_reads_str.append(str(seq_record.seq))\n",
    "            read_counter = Counter(all_reads_str)\n",
    "            for read in read_counter:\n",
    "                outcomes_dict[sample_i][rep][extract_and_match(read,i,rep)] += read_counter[read]\n",
    "            print(samples[rep][i])\n",
    "        except IOError: #this happens when a file is missing\n",
    "            print(\"%s missing\" % samples[rep][i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "##calculate summary data\n",
    "Rep1_Ed_Per = []\n",
    "Rep2_Ed_Per = []\n",
    "Rep3_Ed_Per = []\n",
    "\n",
    "Rep1_No_Region_Per = []\n",
    "Rep2_No_Region_Per = []\n",
    "Rep3_No_Region_Per = []\n",
    "\n",
    "Rep1_No_Site_Per = []\n",
    "Rep2_No_Site_Per = []\n",
    "Rep3_No_Site_Per = []\n",
    "\n",
    "for i in samples.index:\n",
    "    try:\n",
    "        Rep1_Ed_Per.append((float(outcomes_dict[i]['Rep1']['edited']) / (outcomes_dict[i]['Rep1']['edited']+outcomes_dict[i]['Rep1']['wt']))*100)\n",
    "    except ZeroDivisionError:\n",
    "        Rep1_Ed_Per.append(\"div0\")\n",
    "    try:\n",
    "        Rep2_Ed_Per.append((float(outcomes_dict[i]['Rep2']['edited']) / (outcomes_dict[i]['Rep2']['edited']+outcomes_dict[i]['Rep2']['wt']))*100)\n",
    "    except ZeroDivisionError:\n",
    "        Rep2_Ed_Per.append(\"div0\")\n",
    "    try:\n",
    "        Rep3_Ed_Per.append((float(outcomes_dict[i]['Rep3']['edited']) / (outcomes_dict[i]['Rep3']['edited']+outcomes_dict[i]['Rep3']['wt']))*100)\n",
    "    except ZeroDivisionError:\n",
    "        Rep3_Ed_Per.append(\"div0\")\n",
    "    \n",
    "    try:\n",
    "        Rep1_No_Region_Per.append((float(outcomes_dict[i]['Rep1']['unmatched_region']) / (outcomes_dict[i]['Rep1']['edited']+outcomes_dict[i]['Rep1']['wt']+outcomes_dict[i]['Rep1']['unmatched_region']+outcomes_dict[i]['Rep1']['unmatched_edit_nt']))*100)\n",
    "    except ZeroDivisionError:\n",
    "        Rep1_No_Region_Per.append(\"div0\")\n",
    "    try:\n",
    "        Rep2_No_Region_Per.append((float(outcomes_dict[i]['Rep2']['unmatched_region']) / (outcomes_dict[i]['Rep2']['edited']+outcomes_dict[i]['Rep2']['wt']+outcomes_dict[i]['Rep2']['unmatched_region']+outcomes_dict[i]['Rep2']['unmatched_edit_nt']))*100)\n",
    "    except ZeroDivisionError:\n",
    "        Rep2_No_Region_Per.append(\"div0\")\n",
    "    try:\n",
    "        Rep3_No_Region_Per.append((float(outcomes_dict[i]['Rep3']['unmatched_region']) / (outcomes_dict[i]['Rep3']['edited']+outcomes_dict[i]['Rep3']['wt']+outcomes_dict[i]['Rep3']['unmatched_region']+outcomes_dict[i]['Rep3']['unmatched_edit_nt']))*100)\n",
    "    except ZeroDivisionError:\n",
    "        Rep3_No_Region_Per.append(\"div0\")\n",
    "    \n",
    "    try:\n",
    "        Rep1_No_Site_Per.append((float(outcomes_dict[i]['Rep1']['unmatched_edit_nt']) / (outcomes_dict[i]['Rep1']['edited']+outcomes_dict[i]['Rep1']['wt']+outcomes_dict[i]['Rep1']['unmatched_region']+outcomes_dict[i]['Rep1']['unmatched_edit_nt']))*100)\n",
    "    except ZeroDivisionError:\n",
    "        Rep1_No_Site_Per.append(\"div0\")\n",
    "    try:    \n",
    "        Rep2_No_Site_Per.append((float(outcomes_dict[i]['Rep2']['unmatched_edit_nt']) / (outcomes_dict[i]['Rep2']['edited']+outcomes_dict[i]['Rep2']['wt']+outcomes_dict[i]['Rep2']['unmatched_region']+outcomes_dict[i]['Rep2']['unmatched_edit_nt']))*100)\n",
    "    except ZeroDivisionError:\n",
    "        Rep2_No_Site_Per.append(\"div0\")\n",
    "    try:\n",
    "        Rep3_No_Site_Per.append((float(outcomes_dict[i]['Rep3']['unmatched_edit_nt']) / (outcomes_dict[i]['Rep3']['edited']+outcomes_dict[i]['Rep3']['wt']+outcomes_dict[i]['Rep3']['unmatched_region']+outcomes_dict[i]['Rep3']['unmatched_edit_nt']))*100)\n",
    "    except ZeroDivisionError:\n",
    "        Rep3_No_Site_Per.append(\"div0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_updated = samples.assign(Rep1_Edit_Percent=Rep1_Ed_Per, \n",
    "                                 Rep2_Edit_Percent=Rep2_Ed_Per, \n",
    "                                 Rep3_Edit_Percent=Rep3_Ed_Per,\n",
    "                                 _R1_No_Region_Match_Percent=Rep1_No_Region_Per,\n",
    "                                 _R2_No_Region_Match_Percent=Rep2_No_Region_Per,\n",
    "                                 _R3_No_Region_Match_Percent=Rep3_No_Region_Per,\n",
    "                                 __R1_No_Site_Match_Percent=Rep1_No_Site_Per,\n",
    "                                 __R2_No_Site_Match_Percent=Rep2_No_Site_Per,\n",
    "                                 __R3_No_Site_Match_Percent=Rep3_No_Site_Per)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phage</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Plasmid</th>\n",
       "      <th>Direction</th>\n",
       "      <th>edit name</th>\n",
       "      <th>Genome Position</th>\n",
       "      <th>WT nt</th>\n",
       "      <th>Edited nt</th>\n",
       "      <th>L_inside</th>\n",
       "      <th>R_inside</th>\n",
       "      <th>...</th>\n",
       "      <th>Rep3</th>\n",
       "      <th>Rep1_Edit_Percent</th>\n",
       "      <th>Rep2_Edit_Percent</th>\n",
       "      <th>Rep3_Edit_Percent</th>\n",
       "      <th>_R1_No_Region_Match_Percent</th>\n",
       "      <th>_R2_No_Region_Match_Percent</th>\n",
       "      <th>_R3_No_Region_Match_Percent</th>\n",
       "      <th>__R1_No_Site_Match_Percent</th>\n",
       "      <th>__R2_No_Site_Match_Percent</th>\n",
       "      <th>__R3_No_Site_Match_Percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lambda</td>\n",
       "      <td>L</td>\n",
       "      <td>pSBK.147</td>\n",
       "      <td>F</td>\n",
       "      <td>L_TAA_TGA</td>\n",
       "      <td>14126</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>CTTTCGCAGT</td>\n",
       "      <td>AATCCCATGA</td>\n",
       "      <td>...</td>\n",
       "      <td>msCF_02_134</td>\n",
       "      <td>27.062720</td>\n",
       "      <td>26.115981</td>\n",
       "      <td>20.278254</td>\n",
       "      <td>1.406785</td>\n",
       "      <td>2.480698</td>\n",
       "      <td>3.034029</td>\n",
       "      <td>0.097863</td>\n",
       "      <td>0.067915</td>\n",
       "      <td>0.157278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Lambda</td>\n",
       "      <td>L</td>\n",
       "      <td>pSBK.148</td>\n",
       "      <td>R</td>\n",
       "      <td>L_TAA_TGA</td>\n",
       "      <td>14126</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>CTTTCGCAGT</td>\n",
       "      <td>AATCCCATGA</td>\n",
       "      <td>...</td>\n",
       "      <td>msCF_02_135</td>\n",
       "      <td>21.116465</td>\n",
       "      <td>25.431555</td>\n",
       "      <td>15.916106</td>\n",
       "      <td>0.899292</td>\n",
       "      <td>13.225325</td>\n",
       "      <td>1.633095</td>\n",
       "      <td>0.058438</td>\n",
       "      <td>0.047112</td>\n",
       "      <td>0.172252</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Phage Gene   Plasmid Direction  edit name  Genome Position WT nt  \\\n",
       "0  Lambda    L  pSBK.147         F  L_TAA_TGA            14126     A   \n",
       "1  Lambda    L  pSBK.148         R  L_TAA_TGA            14126     A   \n",
       "\n",
       "  Edited nt    L_inside    R_inside  ...         Rep3 Rep1_Edit_Percent  \\\n",
       "0         G  CTTTCGCAGT  AATCCCATGA  ...  msCF_02_134         27.062720   \n",
       "1         G  CTTTCGCAGT  AATCCCATGA  ...  msCF_02_135         21.116465   \n",
       "\n",
       "  Rep2_Edit_Percent Rep3_Edit_Percent _R1_No_Region_Match_Percent  \\\n",
       "0         26.115981         20.278254                    1.406785   \n",
       "1         25.431555         15.916106                    0.899292   \n",
       "\n",
       "   _R2_No_Region_Match_Percent  _R3_No_Region_Match_Percent  \\\n",
       "0                     2.480698                     3.034029   \n",
       "1                    13.225325                     1.633095   \n",
       "\n",
       "   __R1_No_Site_Match_Percent  __R2_No_Site_Match_Percent  \\\n",
       "0                    0.097863                    0.067915   \n",
       "1                    0.058438                    0.047112   \n",
       "\n",
       "   __R3_No_Site_Match_Percent  \n",
       "0                    0.157278  \n",
       "1                    0.172252  \n",
       "\n",
       "[2 rows x 24 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_updated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "##output\n",
    "samples_updated.to_excel(\"editing_analysis_demo_output.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
